web/source/kmwembedded.ts

```diff
@@ -9,9 +12,173 @@
+  VisualKeyboard.prototype.createKeyTip = function(this: VisualKeyboard) {
+    if(com.keyman.singleton.util.device.formFactor == 'phone') {
+      this.keytip = {key:null, state:false};
+    }
+  };
+
+  /**
+   * Adjust the absolute height of each keyboard element after a rotation - modified for KMEI, 13/11/14
+   *    
+   **/      
+  VisualKeyboard.prototype.adjustHeights=function(this: VisualKeyboard): boolean {
+    let keyman = com.keyman.singleton;
+    let osk = keyman.osk;
+    let _Box = osk._Box;
+    let util = keyman.util;
+    let device = util.device;
+
+    if(!_Box || !this.kbdDiv || !this.kbdDiv.firstChild || !this.kbdDiv.firstChild.firstChild.childNodes) {
+      return false;
+    }
+
+    var layers=_Box.firstChild.firstChild.childNodes,
+        nRows=layers[0].childNodes.length,
+        oskHeight=osk.getHeight(),
+        rowHeight=Math.floor(oskHeight/nRows),
+        nLayer,nRow,rs,keys,nKeys,nKey,key,ks,j,pad=4,fs=1.0;
+
+    if(device.OS == 'Android' && 'devicePixelRatio' in window) {
+      rowHeight = rowHeight/window.devicePixelRatio;
+    }
+
+    oskHeight=nRows*rowHeight;
+
+    var b: HTMLElement = osk._Box, bs=b.style;
+    bs.height=bs.maxHeight=(oskHeight+3)+'px';
+    b = <HTMLElement> b.firstChild.firstChild;
+    bs=b.style;
+    bs.height=bs.maxHeight=(oskHeight+3)+'px';
+    pad = Math.round(0.15*rowHeight);
+
+    var resizeLabels=(device.OS == 'iOS' && device.formFactor == 'phone' && util.landscapeView());
+ 
+    for(nLayer=0;nLayer<layers.length; nLayer++) {
+      // Check the heights of each row, in case different layers have different row counts.
+      nRows=layers[nLayer].childNodes.length;
+      (<HTMLElement> layers[nLayer]).style.height=(oskHeight+3)+'px';
+
+      for(nRow=0; nRow<nRows; nRow++) {
+        rs=(<HTMLElement> layers[nLayer].childNodes[nRow]).style;
+        rs.bottom=(nRows-nRow-1)*rowHeight+'px';
+        rs.maxHeight=rs.height=rowHeight+'px';
+        keys=layers[nLayer].childNodes[nRow].childNodes;
+        nKeys=keys.length;
+        for(nKey=0;nKey<nKeys;nKey++) {
+          key=keys[nKey];
+          // Must set the height of the text DIV, not the label (if any)
+          for(j=0;j<key.childNodes.length;j++) {
+            if(util.hasClass(key.childNodes[j],'kmw-key')) {
+              break;
+            }
+          }
+          ks=key.childNodes[j].style;
+          ks.bottom=rs.bottom;
+          ks.height=ks.minHeight=(rowHeight-pad)+'px';
+                          
+          // Rescale keycap labels on iPhone (iOS 7)
+          if(resizeLabels && (j > 0)) key.childNodes[0].style.fontSize='6px';
+        }
+      }
+    }
+
+    return true;
+  };
+}
+
 (function() {
   // Declare KeymanWeb and related objects
-  var keymanweb=window['keyman'],osk=keymanweb['osk'],util=keymanweb['util'],device=util.device;
+  var keymanweb=window['keyman'], osk: com.keyman.osk.OSKManager = keymanweb['osk'],util=keymanweb['util'],device=util.device;
```

Comment:
Would there be less changes if you had this?
```
osk: com.keyman.osk.OSKManager = keymanweb['osk'].vkbd, ...
```

(_Load() might need to move)

Reply:
Maybe so, but this is far more consistent with everything else.

* All `keyman.osk` API functions reside on the `OSKManager` (`keymanweb['osk']`), not on `.vkbd.`
* `KeymanBase` refers to the `OSKManager` through the field `osk`.

Finally, making this suggestion actually impossible to implement:
* `keyman.osk`'s a singleton, while `osk.vkbd` will be destroyed and recreated each time a new keyboard is loaded.
  - After a keyboard change, the 'global' variable would be invalidated.